File: tool/uapicodegen/uapicodegen/uapi/internal/SelectStructFields.py
from typing import TYPE_CHECKING, cast

from ..internal.types.UArray import UArray
from ..internal.types.UFn import UFn
from ..internal.types.UObject import UObject
from ..internal.types.UStruct import UStruct
from ..internal.types.UUnion import UUnion

if TYPE_CHECKING:
    from ..internal.types.UTypeDeclaration import UTypeDeclaration


def select_struct_fields(type_declaration: 'UTypeDeclaration', value: object,
                         selected_struct_fields: dict[str, object]) -> object:
    type_declaration_type = type_declaration.type
    type_declaration_type_params = type_declaration.type_parameters

    if isinstance(type_declaration_type, UStruct):
        fields = type_declaration_type.fields
        struct_name = type_declaration_type.name
        selected_fields = cast(
            list[str] | None, selected_struct_fields.get(struct_name))
        value_as_map = cast(dict[str, object], value)
        final_map = {}

        for field_name, field_value in value_as_map.items():
            if selected_fields is None or field_name in selected_fields:
                field = fields[field_name]
                field_type_declaration = field.type_declaration
                value_with_selected_fields = select_struct_fields(field_type_declaration, field_value,
                                                                  selected_struct_fields)

                final_map[field_name] = value_with_selected_fields

        return final_map
    elif isinstance(type_declaration_type, UFn):
        value_as_map = cast(dict[str, object], value)
        union_tag, union_data = cast(
            tuple[str, dict[str, object]], next(iter(value_as_map.items())))

        fn_name = type_declaration_type.name
        fn_call = type_declaration_type.call
        fn_call_tags = fn_call.tags

        arg_struct_reference = fn_call_tags[union_tag]
        selected_fields = cast(
            list[str] | None, selected_struct_fields.get(fn_name))
        final_map = {}

        for field_name, field_value in union_data.items():
            if selected_fields is None or field_name in selected_fields:
                field = arg_struct_reference.fields[field_name]
                value_with_selected_fields = select_struct_fields(field.type_declaration, field_value,
                                                                  selected_struct_fields)

                final_map[field_name] = value_with_selected_fields

        return {union_tag: final_map}
    elif isinstance(type_declaration_type, UUnion):
        value_as_map = cast(dict[str, object], value)
        union_tag, union_data = cast(
            tuple[str, dict[str, object]], next(iter(value_as_map.items())))

        union_tags = type_declaration_type.tags
        union_struct_reference = union_tags[union_tag]
        union_struct_ref_fields = union_struct_reference.fields
        default_tags_to_fields = {}

        for tag, union_struct in union_tags.items():
            field_names = list(union_struct.fields.keys())
            default_tags_to_fields[tag] = field_names

        union_selected_fields = cast(dict[str, object], selected_struct_fields.get(
            type_declaration_type.name, default_tags_to_fields))
        this_union_tag_selected_fields_default = default_tags_to_fields.get(
            union_tag)
        selected_fields = cast(list[str] | None, union_selected_fields.get(
            union_tag, this_union_tag_selected_fields_default))

        final_map = {}
        for field_name, field_value in union_data.items():
            if selected_fields is None or field_name in selected_fields:
                field = union_struct_ref_fields[field_name]
                value_with_selected_fields = select_struct_fields(field.type_declaration, field_value,
                                                                  selected_struct_fields)

                final_map[field_name] = value_with_selected_fields

        return {union_tag: final_map}
    elif isinstance(type_declaration_type, UObject):
        nested_type_declaration = type_declaration_type_params[0]
        value_as_map = cast(dict[str, object], value)

        final_map = {}
        for key, value in value_as_map.items():
            value_with_selected_fields = select_struct_fields(nested_type_declaration, value,
                                                              selected_struct_fields)

            final_map[key] = value_with_selected_fields

        return final_map
    elif isinstance(type_declaration_type, UArray):
        nested_type = type_declaration_type_params[0]
        value_as_list = cast(list[str], value)

        final_list = []
        for entry in value_as_list:
            value_with_selected_fields = select_struct_fields(
                nested_type, entry, selected_struct_fields)
            final_list.append(value_with_selected_fields)

        return final_list
    else:
        return value

File: lib/java/src/main/java/io/github/msgpact/internal/SelectStructFields.java
package io.github.msgpact.internal;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import io.github.msgpact.internal.types.VArray;
import io.github.msgpact.internal.types.VFieldDeclaration;
import io.github.msgpact.internal.types.VFn;
import io.github.msgpact.internal.types.VObject;
import io.github.msgpact.internal.types.VStruct;
import io.github.msgpact.internal.types.VType;
import io.github.msgpact.internal.types.VTypeDeclaration;
import io.github.msgpact.internal.types.VUnion;

public class SelectStructFields {
    static Object selectStructFields(VTypeDeclaration typeDeclaration, Object value,
            Map<String, Object> selectedStructFields) {
        final VType typeDeclarationType = typeDeclaration.type;
        final List<VTypeDeclaration> typeDeclarationTypeParams = typeDeclaration.typeParameters;

        if (typeDeclarationType instanceof final VStruct s) {
            final Map<String, VFieldDeclaration> fields = s.fields;
            final String structName = s.name;
            final var selectedFields = (List<String>) selectedStructFields.get(structName);
            final var valueAsMap = (Map<String, Object>) value;
            final var finalMap = new HashMap<>();

            for (final var entry : valueAsMap.entrySet()) {
                final var fieldName = entry.getKey();
                if (selectedFields == null || selectedFields.contains(fieldName)) {
                    final var field = fields.get(fieldName);
                    final VTypeDeclaration fieldTypeDeclaration = field.typeDeclaration;
                    final var valueWithSelectedFields = selectStructFields(fieldTypeDeclaration, entry.getValue(),
                            selectedStructFields);

                    finalMap.put(entry.getKey(), valueWithSelectedFields);
                }
            }

            return finalMap;
        } else if (typeDeclarationType instanceof final VFn f) {
            final var valueAsMap = (Map<String, Object>) value;
            final Map.Entry<String, Object> uEntry = valueAsMap.entrySet().stream().findAny().get();
            final var unionTag = uEntry.getKey();
            final var unionData = (Map<String, Object>) uEntry.getValue();

            final String fnName = f.name;
            final VUnion fnCall = f.call;
            final Map<String, VStruct> fnCallTags = fnCall.tags;

            final var argStructReference = fnCallTags.get(unionTag);
            final var selectedFields = (List<String>) selectedStructFields.get(fnName);
            final var finalMap = new HashMap<>();

            for (final var entry : unionData.entrySet()) {
                final var fieldName = entry.getKey();
                if (selectedFields == null || selectedFields.contains(fieldName)) {
                    final var field = argStructReference.fields.get(fieldName);
                    final var valueWithSelectedFields = selectStructFields(field.typeDeclaration, entry.getValue(),
                            selectedStructFields);

                    finalMap.put(entry.getKey(), valueWithSelectedFields);
                }
            }

            return Map.of(uEntry.getKey(), finalMap);
        } else if (typeDeclarationType instanceof final VUnion u) {
            final var valueAsMap = (Map<String, Object>) value;
            final var uEntry = valueAsMap.entrySet().stream().findAny().get();
            final var unionTag = uEntry.getKey();
            final var unionData = (Map<String, Object>) uEntry.getValue();

            final Map<String, VStruct> unionTags = u.tags;
            final var unionStructReference = unionTags.get(unionTag);
            final var unionStructRefFields = unionStructReference.fields;
            final var defaultTagsToFields = new HashMap<String, List<String>>();

            for (final var entry : unionTags.entrySet()) {
                final var unionStruct = entry.getValue();
                final var unionStructFields = unionStruct.fields;
                final var fieldNames = unionStructFields.keySet().stream().toList();
                defaultTagsToFields.put(entry.getKey(), fieldNames);
            }

            final var unionSelectedFields = (Map<String, Object>) selectedStructFields.getOrDefault(u.name,
                    defaultTagsToFields);
            final var thisUnionTagSelectedFieldsDefault = defaultTagsToFields.get(unionTag);
            final var selectedFields = (List<String>) unionSelectedFields.getOrDefault(unionTag,
                    thisUnionTagSelectedFieldsDefault);

            final var finalMap = new HashMap<>();
            for (final var entry : unionData.entrySet()) {
                final var fieldName = entry.getKey();
                if (selectedFields == null || selectedFields.contains(fieldName)) {
                    final var field = unionStructRefFields.get(fieldName);
                    final var valueWithSelectedFields = selectStructFields(field.typeDeclaration, entry.getValue(),
                            selectedStructFields);
                    finalMap.put(entry.getKey(), valueWithSelectedFields);
                }
            }

            return Map.of(uEntry.getKey(), finalMap);
        } else if (typeDeclarationType instanceof final VObject o) {
            final var nestedTypeDeclaration = typeDeclarationTypeParams.get(0);
            final var valueAsMap = (Map<String, Object>) value;

            final var finalMap = new HashMap<>();
            for (final var entry : valueAsMap.entrySet()) {
                final var valueWithSelectedFields = selectStructFields(nestedTypeDeclaration, entry.getValue(),
                        selectedStructFields);
                finalMap.put(entry.getKey(), valueWithSelectedFields);
            }

            return finalMap;
        } else if (typeDeclarationType instanceof final VArray a) {
            final var nestedType = typeDeclarationTypeParams.get(0);
            final var valueAsList = (List<Object>) value;

            final var finalList = new ArrayList<>();
            for (final var entry : valueAsList) {
                final var valueWithSelectedFields = selectStructFields(nestedType, entry, selectedStructFields);
                finalList.add(valueWithSelectedFields);
            }

            return finalList;
        } else {
            return value;
        }
    }
}
